docs/input processing.md

```diff
@@ -0,0 +1,40 @@
+# Input processing
+The input process is about processing any input into our internal model representation `CommonInputModel`. This most likely have to be done using different methods based on which kind of input it is.
+
+As of now two inputs are supported:
+- JSON Schema Draft 7
+- AsyncAPI version 2.0.0
+
+## Internal model representation
+
+<img src="./assets/class_diagram.png"
+     alt="Class diagram" />
+
+As seen on the class diagram the `InputProcessor` is our main point of entry for processing input data. It uses the defined other input processors (`AsyncAPIInputProcessor`, `JsonSchemaInputProcessor`, ...) by calling `shouldProcess` function of each and if the function returns true processing it using the `process` function. If no processes returns true it defaults to `JsonSchemaInputProcessor`. 
+
+The `process` function are expected to return `CommonInputModel` which is a wrapper for the core data representation of `CommonModel`. This is done to ensure we can return multiple models for any input to allow for references, inheritance etc. As said the core internal representation of a data model is `CommonModel`. This contains the data definition by using known keywords from JSON Schema, but instead of it representing a validation rules it represent data definitions.
+
+Here are the explanation of the different properties in `CommonModel`:
+- `$id` is the id/name of the model.
+- `type` this is the different types for the model. All types from JSON Schema are used, no custom types are used.
+- `enum` defines the different enums for the model.
+- `items` defines the type for `array` models as `CommonModel`.
+- `properties` defines the properties and its expected types as `CommonModel`.
+- `additionalProperties` are used to define if any extra properties are allowed, also defined as a  `CommonModel`.
+- `$ref` is a reference to another `CommonModel` by using`$id` as a simple string.
+- `required` list of required properties.
+- `extend` list of other `CommonModel`s this model extends, is an array of `$id` strings.
+- `originalSchema` is the actual input which this model represent.
```

Comment:
I'm a bit concerned about the possibility of this becoming out of date earlier than expected. Maintaining this here rather than in code. Is there a way we can move all this to become always up-to-date with code generation? Perhaps documenting code?

Reply:
Hmm, yea you are right. Let me see if jsdoc provides some functionality for us when we extend classes.

Reply:
@smoya updated with link to generated docs.